README.md
# Checkout Finland for WooCommerce

## ⚠️ DEPRECATED

**This plugin has been deprecated**. 

All of its functionality now exists as the rebranded [`paytrail/paytrail-for-woocommerce`](https://github.com/paytrail/paytrail-for-woocommerce) plugin.<br/>
It is advised to upgrade to the new plugin as this one is no longer maintained.

<hr/>

**Required WordPress version:** 4.9

**Required PHP version:** 7.3

**[Checkout Finland](https://www.checkout.fi)** is a payment gateway that offers 20+ payment methods for Finnish customers. NOTICE: This plugin is no longer maintained. Use Paytrail for Woocommerce instead.

The payment gateway provides all the popular payment methods with one simple integration. The provided payment methods include but are not limited to credit cards, online banking and mobile payments.

To use this extension, you need to sign up for a Checkout Finland account. Transaction fees will be charged for every transaction. Transaction cost may vary from merchant to merchant, based on what is agreed upon with Checkout Finland when negotiating your contract. For more information and to register, please visit [our website](https://www.checkout.fi)  (in Finnish only) or contact [asiakaspalvelu@checkout.fi](mailto:asiakaspalvelu@checkout.fi) directly.

We employ the industry's best security practices and tools to maintain bank-level security for merchants and end customers. Checkout Finland is PCI DSS Level I and GDPR compliant. 

Upon checkout, customers are redirected to the Checkout Finland payment service. The customer enters his or her payment information directly into our secure environment so that the web shop never comes into contact with the customers payment data. Once the payment process is complete, customers will redirected back to your store. Tokenization is used to run transactions with stored payment information. No confidential card data is ever stored on your server.

## Installation

### From WordPress plugin directory

Open WordPress Admin panel and go to Plugins -> Add New. Search for [Checkout Finland for WooCommerce](https://wordpress.org/plugins/op-payment-service-for-woocommerce/), click install and then activate. 

After installation go to WooCommerce -> Settings -> Payments and select "Manage" next to Checkout Finland for WooCommerce to review settings.

### Via Composer

1. If you have Composer installed:
- You can use the command line to install the plugin:

```
$ composer require checkout-finland/checkout-finland-for-woocommerce
```
- Or you can add the following json to your `composer.json` file:

```json
{
  "require": {
    "checkout-finland/checkout-finland-for-woocommerce": "*"
  }
}
```
2. Activate the plugin.
3. Go to WooCommerce Settings and open Payments tab.
4. Enable Checkout Finland for WooCommerce with the toggle switch.
5. Configure your own Checkout Finland payment service settings.

## Configuration

There are several settings you can configure via the options page.

### Test mode

If test mode is enabled, your store will automatically use Checkout Finland's test credentials.

### Debug log

Enable logging all payment gateway events. Recommended only for debugging purposes as this might save personal data.

### Payment provider selection

Choose whether you want the payment provider selection to happen in the checkout page or in a separate page.

### Merchant ID and Secret key

Your Checkout Finland credentials. You can obtain the keys by registering at our [website](https://www.checkout.fi) (in Finnish only) or contact [asiakaspalvelu@checkout.fi](mailto:asiakaspalvelu@checkout.fi) directly.

### Fallback Country

Setup fallback country to be used if no country is provided from Woocommerce checkout. Use only if billing_country/shipping_country is removed from checkout fields.

## Router

### Callback Routes

Permalinks enabled:

    Success: /op-payment-service/callback/index
    Cancel: /op-payment-service/callback/index

Permalinks disabled:

    Success: /index.php?op-payment-service-route=callback&op-payment-service-action=index
    Cancel: /index.php?op-payment-service-route=callback&op-payment-service-action=index

## Credits

**Contributors:** [Nomafin](https://github.com/nomafin), [villesiltala](https://github.com/villesiltala), [tommireinikainen](https://github.com/tommireinikainen), [onnimonni](https://github.com/onnimonni), [traone](https://github.com/traone), [villepietarinen](https://github.com/villepietarinen) & [loueranta](https://github.com/loueranta)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.7.1] - 2021-12-08
### Added
- Added deprecation notice

## [1.7] - 2021-09-30
### Changed
- PHP SDK updated to 1.5 version
- Removed Guzzle dependency

## [1.6] - 2021-06-09
### Changed
- Rebranding from "OP Payment Service" to "Checkout Finland"

## [1.5.15] - 2021-02-23
### Changed
- Add validation for order tokens when creating scheduled Subscription/MIT payment
- Add better error message if no tokens found
- Change PHP version to 7.4 in Github Action

## [1.5.14] - 2021-02-17
### Changed
- Update minimum PHP version to 7.3
- OP Payment Service PHP-SDK version updated to 1.3.4

## [1.5.13] - 2021-02-16
### Changed
- OP Payment Service PHP-SDK version updated

## [1.5.12] - 2021-02-15
### Changed
- Add more logging to gateway
- Hide the payment method group title and expand the group if only one group is available

### Fixed
- Fix payment method description support
- Fix error handling when no payment provider chosen
- Fix initialising of payment method groups in provider form

## [1.5.11] - 2020-12-11
### Fixed
- Add missing register scripts

## [1.5.10] - 2020-12-11
### Fixed
- Updated dist script

## [1.5.9] - 2020-12-11
### Changed
- Added support for WP 5.6 and WC 4.8

## [1.5.8] - 2020-11-10
### Changed
- Added order locking in redirect/callback processing
- Improved logging

## [1.5.7] - 2020-09-11
### Changed
- Changed callback handling 
- Changed log reference

## [1.5.6] - 2020-09-08
### Changed
- Changed the payment page customization texts
- Add callback mode as parameter to gateway and move sleep to gateway, add order ids to log messages

### Fixed
- Fixed 'Highlighted payment method' custom color issue

## [1.5.5] - 2020-09-07
### Changed
- Changed to use the order number as a reference for the payment service providers instead of a random number
- Add delay to callback controller to prevent simultaneous processing of payments

## [1.5.3] - 2020-09-01
### Changed
- Add more debug logging to payment gateway processing
- Use scrolIntoView method after clicking payment method group button on checkout

## [1.5.2] - 2020-07-27
### Changed
- Add extra validation to check that wanted_provider is not null, otherwise use capitalized provider name from checkout-provider parameter.
- Change get_item_description function item description for product items to use either product name or product id
- Add composer.json to require guzzlehttp/guzzle version 6.5.4
- Update finnish translations

### Fixed
- Fix issue when using a gift card code after placing an order

## [1.5.1] - 2020-06-29
### Changed
- Add setting for fallback country in payment gateway settings 

## [1.5.0] - 2020-06-16
### Changed
- Add support for WooCommerce Subscriptions plugin
- Add tokenized card MIT payments functionalities
- Add new Helper class
- Add info and error logging for payment requests if debug log enabled
- Updated to the latest PHP-SDK version

## [1.4.1] - 2020-05-27
### Changed
- New added tokenized card automatically set as users default token payment method
- Updated to the latest PHP-SDK version

## [1.4.0] - 2020-05-26
### Changed
- Add support for adding and using tokenized cards as payment methods
- Add tokenized card CIT payments functionalities
- Add Callback controller class and implementation

## [1.3.1] - 2020-04-15
### Fixed
- Fixed the hosted payment method selection

## [1.3.0] - 2020-04-03
### Changed
- New layout for the payment method selection

## [1.2.8] - 2020-01-28
### Changed
- Bump SDK version to composer.lock 

## [1.2.7] - 2020-01-28
### Changed
- Simplified error handling
- Strict comparison to some validations
- Use mb_substr() to truncate the description

## [1.2.6] - 2020-01-08
### Changed
- Updated to the latest PHP-SDK version
- Version bump to plugin.php 

## [1.2.2] - 2020-01-08
### Changed
- Updated to the latest PHP-SDK version

### Added
- Custom title and description to be used on the Payment provider section.

## [1.2.1] - 2019-10-24
### Changed
- Small fixes

## [1.2.0] - 2019-10-22
### Changed
- Rebranding from Checkout Finland to OP Payment Services for WooCommerce

## [1.1.7] - 2019-10-04
### Fixed
- A bug where an error message from the provider list API would throw a fatal error to user.
- Better PHP version checking for plugin activation.
- Better error handling for when payment provider is not selected.

## [1.1.6] - 2019-09-24
### Fixed
- Order items without taxes no longer cause 500 error on checkout. [Fixes: #9](https://github.com/CheckoutFinland/woocommerce-checkout-finland-gateway/issues/9).

## [1.1.5] - 2019-08-28
### Removed
- Removed the manual installation guide from readme. Manual installation is not yet supported.

## [1.1.4] - 2019-08-28
### Added
- Pass the current order object for WooCommerce when creating the return url. This enables displaying the order data on a "thank you" page.

## [1.1.3] - 2019-06-12
### Changed
- Updated correct dependencies for composer/packagist.

## [1.1.2] - 2019-06-4
### Changed
- Use 1.0.0 tag for php-sdk instead of dev-master.

## [1.1.1] - 2019-05-16
### Added
- A feature that truncates the product description to a maximum of 1000 characaters when sending it to the Checkout Finland API.

## [1.1.0-beta] - 2019-04-24
### Added
- A feature to add a rounding row to the order in case the roundings are off by a few cents.

### Changed
- Better handling for the situation where WooCommerce is not activated when the user tries to activate the gateway.

## [1.0.7-beta] - 2019-04-15
### Fixed
- Correct version number to be passed to the API.
- Another small stylistical fix for the payment provider logo grid.

## [1.0.6-beta] - 2019-04-09
### Fixed
- A small stylistical fix for the payment provider logo grid.

## [1.0.5-beta] - 2019-04-09
### Fixed

plugin.php
<?php
/**
 * Plugin Name: Checkout Finland for WooCommerce
 * Plugin URI: https://github.com/CheckoutFinland/checkout-finland-for-woocommerce
 * Description: Notice: This plugin is no longer maintained. Use Paytrail for WooCommerce instead. Checkout Finland is a payment gateway that offers 20+ payment methods for Finnish customers.
 * Version: 1.7.1
 * Requires at least: 4.9
 * Tested up to: 5.8
 * Requires PHP: 7.3
 * WC requires at least: 3.0
 * WC tested up to: 5.7
 * Author: Checkout Finland
 * Author URI: https://www.checkout.fi/
 * Text Domain: op-payment-service-woocommerce
 * Domain Path: /languages
 * License: MIT
 * License URI: https://opensource.org/licenses/MIT
 * Copyright: Checkout Finland
 */

namespace OpMerchantServices\WooCommercePaymentGateway;

// Ensure that the file is being run within the WordPress context.
if ( ! defined( 'ABSPATH' ) ) {
    die;
}

/**
 * The main plugin class
 */
final class Plugin {

    /**
     * WooCommerce payment gateway ID.
     */
    public const GATEWAY_ID = 'checkout_finland';

    /**
     * Merchant ID for the test mode.
     */
    public const TEST_MERCHANT_ID = 375917;

    /**
     * Secret key for the test mode.
     */
    public const TEST_SECRET_KEY = 'SAIPPUAKAUPPIAS';

    /**
     * The URL from which the method description should be fetched.
     */
    public const METHOD_INFO_URL = 'https://cdn2.hubspot.net/hubfs/2610868/ext-media/op-psp-service-info.json';

    /**
     * The URL of the payment method icon
     */
    public const ICON_URL = 'https://cdn2.hubspot.net/hubfs/2610868/ext-media/op-psp-master-logo.svg';

    public const PAYMENT_METHOD_IMG_URL = 'https://payment.checkout.fi/static/img/payment-methods';

    public const BASE_URL = 'op-payment-service/';

    public const ADD_CARD_REDIRECT_SUCCESS_URL = 'card-success';

    public const ADD_CARD_REDIRECT_CANCEL_URL = 'card-cancel';

    public const ADD_CARD_CONTEXT_MY_ACCOUNT = 'my_account';

    public const ADD_CARD_CONTEXT_CHECKOUT= 'checkout';

    public const ADD_CARD_CONTEXT_CHANGE_PAYMENT_METHOD = 'change_payment_method';

    public const CARD_ENDPOINT = 'card';

    public const CALLBACK_URL = 'callback';

    /**
     * Singleton instance.
     *
     * @var Plugin
     */
    private static $instance;

    /**
     * Plugin version.
     *
     * @var string
     */
    public static $version;

    /**
     * Plugin directory.
     *
     * @var string
     */
    protected $plugin_dir;

    /**
     * Plugin directory URL.
     *
     * @var string
     */
    protected $plugin_dir_url;

    /**
     * Container array for possible initialization errors.
     *
     * @var array
     */
    protected $errors = [];

    /**
     * Plugin info
     *
     * @var array
     */
    protected $plugin_info = [
        'Plugin Name',
        'Plugin URI',
        'Description',
        'Version',
        'Author',
        'Author URI',
        'Text Domain',
        'Domain Path',
    ];

    /**
     * Constructor function
     */
    protected function __construct() {
        $this->plugin_dir     = __DIR__;
        $this->plugin_dir_url = plugin_dir_url( __FILE__ );
        $this->plugin_info    = array_combine( $this->plugin_info, get_file_data( __FILE__, $this->plugin_info ) );

        self::$version = $this->plugin_info['Version'];

        // Load the plugin textdomain.
        load_plugin_textdomain( 'op-payment-service-woocommerce', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );
        // Register admin notice about plugin status
        add_action( 'admin_notices', [$this, 'admin_notice_maintain'] );
        // Register customizations
        add_action( 'customize_register', [ $this, 'checkout_customizations' ] );
        // Add custom styles
        add_action( 'wp_head', [ $this, 'op_checkout_customize_css' ] );
        // Enable WP Dashicons on frontend
        add_action( 'wp_enqueue_scripts', function() {
            wp_enqueue_style( 'dashicons' );
        } );
    }

    /**
     * Print custom styles
     */
    public function op_checkout_customize_css() {
        ?>
            <style type="text/css">
                .provider-group {
                    background-color: <?php echo get_theme_mod('op_group_background', '#ebebeb'); ?> !important;
                    color: <?php echo get_theme_mod('op_group_text', '#515151'); ?> !important;
                }
                .provider-group.selected {
                    background-color: <?php echo get_theme_mod('op_group_highlighted_background', '#33798d'); ?> !important;
                    color: <?php echo get_theme_mod('op_group_highlighted_text', '#ffffff'); ?> !important;
                }
                .provider-group.selected div {
                    color: <?php echo get_theme_mod('op_group_highlighted_text', '#ffffff'); ?> !important;
                }
                .provider-group:hover {
                    background-color: <?php echo get_theme_mod('op_group_hover_background', '#d0d0d0'); ?> !important;
                    color: <?php echo get_theme_mod('op_group_hover_text', '#515151'); ?> !important;
                }
                .provider-group.selected:hover {
                    background-color: <?php echo get_theme_mod('op_group_highlighted_background', '#33798d'); ?> !important;
                    color: <?php echo get_theme_mod('op_group_highlighted_text', '#ffffff'); ?> !important;
                }
                .woocommerce-checkout #payment .op-payment-service-woocommerce-payment-fields--list-item--input:checked+.op-payment-service-woocommerce-payment-fields--list-item--wrapper, .woocommerce-checkout #payment .op-payment-service-woocommerce-payment-fields--list-item:hover .op-payment-service-woocommerce-payment-fields--list-item--wrapper {                    
                    border: 2px solid <?php echo get_theme_mod('op_method_highlighted', '#33798d'); ?> !important;
                }
                .woocommerce-checkout #payment ul.payment_methods li.op-payment-service-woocommerce-payment-fields--list-item .op-payment-service-woocommerce-payment-fields--list-item--wrapper:hover {
                    border: 2px solid <?php echo get_theme_mod('op_method_hover', '#5399ad'); ?> !important;
                }
            </style>
        <?php
    }

    public function admin_notice_maintain() {
        $allowed_html = array(
            'a'      => array(
                'href'  => array(),
                'title' => array(),
                'target' => array(),
            )
        );
        $class = 'notice notice-error is-dismissible';
        $message = __( 'Notice: Checkout Finland for Woocommerce is deprecated. To continue using the payment service, install this plugin: <a href="https://wordpress.org/plugins/paytrail-for-woocommerce/" target="_blank">Paytrail for Woocommerce</a>.', 'op-payment-service-woocommerce' );
        printf( '<div class="%1$s"><p>%2$s</p></div>', esc_attr( $class ), wp_kses( $message,$allowed_html ) ); 
    }
    

    /**
     * Customizer options
     */
    public function checkout_customizations( $wp_customize ) {
        // Settings
        $wp_customize->add_setting( 'op_group_background' , array(
            'default'   => '#ebebeb',
            'transport' => 'refresh',
        ) );
        $wp_customize->add_setting( 'op_group_text' , array(
            'default'   => '#515151',
            'transport' => 'refresh',
        ) );
        $wp_customize->add_setting( 'op_group_highlighted_background' , array(
            'default'   => '#33798d',
            'transport' => 'refresh',
        ) );
        $wp_customize->add_setting( 'op_group_highlighted_text' , array(
            'default'   => '#ffffff',
            'transport' => 'refresh',
        ) );
        $wp_customize->add_setting( 'op_group_hover_background' , array(
            'default'   => '#d0d0d0',
            'transport' => 'refresh',
        ) );
        $wp_customize->add_setting( 'op_group_hover_text' , array(
            'default'   => '#313131',
            'transport' => 'refresh',
        ) );
        $wp_customize->add_setting( 'op_method_highlighted' , array(
            'default'   => '#33798d',
            'transport' => 'refresh',
        ) );
        $wp_customize->add_setting( 'op_method_hover' , array(
            'default'   => '#5399ad',
            'transport' => 'refresh',
        ) );
        // Section
        $wp_customize->add_section( 'op_checkout_customize_section' , array(
            'title'      => __( 'Payment page personalization', 'op-payment-service-woocommerce' ),
            'priority'   => 30,
        ) );
        // Controls
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_bgcolor', array(
            'label'      => __( 'Payment method group background', 'op-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_group_background',
        ) ) );
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_fgcolor', array(
            'label'      => __( 'Payment method group text', 'op-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_group_text',
        ) ) );
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_bgcolor_selected', array(
            'label'      => __( 'Selected payment method group background', 'op-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_group_highlighted_background',
        ) ) );
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_fgcolor_selected', array(
            'label'      => __( 'Selected payment method group text', 'op-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_group_highlighted_text',
        ) ) );
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_bgcolor_hover', array(
            'label'      => __( 'Payment method group background hover', 'op-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_group_hover_background',
        ) ) );
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_fgcolor_hover', array(
            'label'      => __( 'Payment method group text hover', 'op hover-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_group_hover_text',
        ) ) );
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_bordercolor_selected', array(
            'label'      => __( 'Selected payment method', 'op-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_method_highlighted',
        ) ) );
        $wp_customize->add_control( new \WP_Customize_Color_Control( $wp_customize, 'op_bordercolor_hover', array(
            'label'      => __( 'Payment method hover', 'op-payment-service-woocommerce' ),
            'section'    => 'op_checkout_customize_section',
            'settings'   => 'op_method_hover',
        ) ) );
    }

    /**
     * Singleton instance getter function
     *
     * @return Plugin
     */
    public static function instance() {
        if ( is_null( self::$instance ) ) {
            // Construct the object.
            self::$instance = new self();

            // Run initialization checks. If any of the checks
            // fails, interrupt the execution.
            if ( ! self::$instance->initialization_checks() ) {
                return;
            }

            // Check if Composer has been initialized in this directory.
            // Otherwise we just use global composer autoloading.
            if ( file_exists( __DIR__ . '/vendor/autoload.php' ) ) {
                require_once __DIR__ . '/vendor/autoload.php';
            }

            // Create new instance of Router class
            new Router();

            // Add the gateway class to WooCommerce.
            add_filter( 'woocommerce_payment_gateways', function( $gateways ) {
                $gateways[] = Gateway::CLASS;

                return $gateways;
            });
        }

        return self::$instance;
    }

    /**
     * Run checks for plugin requirements.
     *
     * Returns false if checks failed.
     *
     * @return bool
     */
    protected function initialization_checks() {
        $errors = [];

        $errors[] = self::check_php_version();
        $errors[] = self::check_woocommerce_active_status();
        $errors[] = self::check_woocommerce_version();

        $errors = array_filter( $errors );

        if ( ! empty( $errors ) ) {
            add_action( 'admin_notices', function() use ( $errors ) {
                echo '<div class="notice notice-error">';
                array_walk( $errors, 'esc_html_e' );
                echo '</div>';
            });

            return false;
        }
        else {
            return true;
        }
    }

    /**
     * Checks to run on plugin activation
     *
     * @return void
     */
    public static function activation_check() {
        $checks = [
            'check_php_version',
            'check_woocommerce_active_status',
            'check_woocommerce_version',
        ];

        array_walk( $checks, function( $check ) {
            $error = call_user_func( __CLASS__ . '::' . $check );

            if ( $error ) {
                wp_die( esc_html( $error ) );
            }
        });
    }

    /**
     * Ensure that the PHP version is at least 7.0.0.
     *
     * @return string|null
     */
    public static function check_php_version() : ?string {
        if ( ! version_compare( PHP_VERSION, '7.1.0', '>=' ) ) {
            return sprintf(
                // translators: The placeholder contains the current PHP version.
                esc_html__( 'Checkout Finland payment gateway plugin requires a PHP version of at least 7.1. You are currently running version %1$s.', 'op-payment-service-woocommerce' ),
                esc_html( PHP_VERSION )
            );
        }

        return null;
    }

    /**
     * Ensure that the WooCommerce plugin is active.
     *
     * @return string|null
     */
    public static function check_woocommerce_active_status() : ?string {
        if ( ! class_exists( '\WC_Payment_Gateway' ) ) {
            return esc_html__( 'Checkout Finland payment gateway plugin requires WooCommerce to be activated.', 'op-payment-service-woocommerce' );
        }

        return null;
    }

    /**
     * Ensure that we have at least version 3.5 of the WooCommerce plugin.
     *
     * @return string|null
     */
    public static function check_woocommerce_version() : ?string {
        if (
            defined( 'WOO_COMMERCE_VERSION' ) &&
            version_compare( WOO_COMMERCE_VERSION, '3.5' ) === -1
        ) {
            return esc_html__( 'Checkout Finland gateway plugin requires WooCommerce version of 3.5 or greater.', 'op-payment-service-woocommerce' );
        }

        return null;
    }

    /**
     * Get plugin directory.
     *
     * @return string
     */
    public function get_plugin_dir() : string {
        return $this->plugin_dir;
    }

    /**
     * Get plugin directory URL.
     *
     * @return string
     */
    public function get_plugin_dir_url() : string {
        return $this->plugin_dir_url;
    }

    /**
     * Get plugin info.
     *
     * @return array
     */
    public function get_plugin_info() : array {
        return $this->plugin_info;
    }
}

add_action( 'plugins_loaded', function() {
    Plugin::instance();
});


register_activation_hook( __FILE__, __NAMESPACE__ . '\\Plugin::activation_check' );
